x-pack/elastic-agent/pkg/agent/cmd/common.go

```diff
@@ -71,7 +83,75 @@ func NewCommandWithArgs(args []string, streams *cli.IOStreams) *cobra.Command {
 	if reexec != nil {
 		cmd.AddCommand(reexec)
 	}
+	cmd.PersistentPreRunE = preRunCheck(flags)
 	cmd.Run = run.Run
 
 	return cmd
 }
+
+func preRunCheck(flags *globalFlags) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		commitFilepath := filepath.Join(paths.Home(), commitFile)
+		content, err := ioutil.ReadFile(commitFilepath)
+
+		// file is not present we are at child
+		if os.IsNotExist(err) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		pathConfigFile := flags.Config()
+		rawConfig, err := config.LoadYAML(pathConfigFile)
+		if err != nil {
+			return errors.New(err,
+				fmt.Sprintf("could not read configuration file %s", pathConfigFile),
+				errors.TypeFilesystem,
+				errors.M(errors.MetaKeyPath, pathConfigFile))
+		}
+
+		cfg, err := configuration.NewFromConfig(rawConfig)
+		if err != nil {
+			return errors.New(err,
+				fmt.Sprintf("could not parse configuration file %s", pathConfigFile),
+				errors.TypeFilesystem,
+				errors.M(errors.MetaKeyPath, pathConfigFile))
+		}
+
+		logger, err := logger.NewFromConfig("", cfg.Settings.LoggingConfig)
+		if err != nil {
+			return err
+		}
+
+		reexecPath := filepath.Join(paths.Data(), hashedDirName(content), os.Args[0])
+		rexLogger := logger.Named("reexec")
+		argsOverrides := []string{
+			"--path.data", paths.Data(),
+			"--path.home", filepath.Dir(reexecPath),
+			"--path.config", paths.Config(),
+		}
+		rm := reexec.Manager(rexLogger, reexecPath)
+		rm.ReExec(argsOverrides...)
```

Comment:
I don't think this is really going to work on Windows. being that Windows doesn't support proper re-exec.

Running from Powershell will show this weird behavior, where calling `elastic-agent run -e` will exit back to the shell but the logs keep printing to the terminal.

Running as a service I looks like it would just be an infinite loop. It would keep restarting the service, and the service would then just keep re-exec, because this same binary would be pointed at by the service.

This is a good approach on Unix based systems, but on Windows I don't believe it will work.

What about the symlink idea? That seems like it would work on all OS types, including Windows: https://golang.org/pkg/syscall/?GOOS=windows#CreateSymbolicLink

Reply:
i will need to test it on windows, symlink is ok but needs to be tested as well. we need to create symlink at the build time or provide some activation script, start script or something similar. 
i havent tried but i think that creating a symlink during release would not result in something working when you unpack it locally.